### ISSUE TITLE
CheckIdleResourcesTask.run() throws ClassNotFoundException when redeploying the application in Tomcat

### ISSUE THREAD

**Issue by james-hu** [swaldman/c3p0#66]
When redeploying the application in Tomcat, the thread running CheckIdleResourcesTask.run() cannot be terminated by Tomcat properly because logger classes are already unavailable.

Perhaps should just catch ClassNotFoundException and quit gracefully.

Below is from catalina.out:
```
INFO: Illegal access: this web application instance has been stopped already.  Could not load ch.qos.logback.core.util.ExecutorServiceUtil.  The eventual following stack trace is caused by an error thrown for debugging purposes as well as to attempt to terminate the thread which caused the illegal access, and has no functional impact.
java.lang.IllegalStateException
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1743)
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1701)
	at ch.qos.logback.core.ContextBase.getExecutorService(ContextBase.java:166)
	at ch.qos.logback.classic.turbo.ReconfigureOnChangeFilter.detachReconfigurationToNewThread(ReconfigureOnChangeFilter.java:158)
	at ch.qos.logback.classic.turbo.ReconfigureOnChangeFilter.decide(ReconfigureOnChangeFilter.java:122)
	at ch.qos.logback.classic.spi.TurboFilterList.getTurboFilterChainDecision(TurboFilterList.java:51)
	at ch.qos.logback.classic.LoggerContext.getTurboFilterChainDecision_0_3OrMore(LoggerContext.java:252)
	at ch.qos.logback.classic.Logger.callTurboFilters(Logger.java:772)
	at ch.qos.logback.classic.Logger.isDebugEnabled(Logger.java:490)
	at ch.qos.logback.classic.Logger.isDebugEnabled(Logger.java:486)
	at com.mchange.v2.log.slf4j.Slf4jMLog$Slf4jMLogger.levelLogger(Slf4jMLog.java:166)
	at com.mchange.v2.log.slf4j.Slf4jMLog$Slf4jMLogger.isLoggable(Slf4jMLog.java:332)
	at com.mchange.v2.resourcepool.BasicResourcePool$CheckIdleResourcesTask.run(BasicResourcePool.java:2173)
	at java.util.TimerThread.mainLoop(Timer.java:555)
	at java.util.TimerThread.run(Timer.java:505)

Exception in thread "C3P0PooledConnectionPoolManager[identityToken->2uy8nq9f1ei09tid3uczj|5dd09579]-AdminTaskTimer" java.lang.NoClassDefFoundError: ch/qos/logback/core/util/ExecutorServiceUtil
	at ch.qos.logback.core.ContextBase.getExecutorService(ContextBase.java:166)
	at ch.qos.logback.classic.turbo.ReconfigureOnChangeFilter.detachReconfigurationToNewThread(ReconfigureOnChangeFilter.java:158)
	at ch.qos.logback.classic.turbo.ReconfigureOnChangeFilter.decide(ReconfigureOnChangeFilter.java:122)
	at ch.qos.logback.classic.spi.TurboFilterList.getTurboFilterChainDecision(TurboFilterList.java:51)
	at ch.qos.logback.classic.LoggerContext.getTurboFilterChainDecision_0_3OrMore(LoggerContext.java:252)
	at ch.qos.logback.classic.Logger.callTurboFilters(Logger.java:772)
	at ch.qos.logback.classic.Logger.isDebugEnabled(Logger.java:490)
	at ch.qos.logback.classic.Logger.isDebugEnabled(Logger.java:486)
	at com.mchange.v2.log.slf4j.Slf4jMLog$Slf4jMLogger.levelLogger(Slf4jMLog.java:166)
	at com.mchange.v2.log.slf4j.Slf4jMLog$Slf4jMLogger.isLoggable(Slf4jMLog.java:332)
	at com.mchange.v2.resourcepool.BasicResourcePool$CheckIdleResourcesTask.run(BasicResourcePool.java:2173)
	at java.util.TimerThread.mainLoop(Timer.java:555)
	at java.util.TimerThread.run(Timer.java:505)
Caused by: java.lang.ClassNotFoundException: ch.qos.logback.core.util.ExecutorServiceUtil
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1858)
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1701)
	... 13 more
```

**Comment by swaldman** [comment]
hi,

are you close()ing your DataSource prior to web-app restart (e.g. in a ServletContextListener)?

thanks,
    steve

**Comment by james-hu** [comment]
Hi Steve,
Thanks for the prompt reply. New finding to share: I created two pools and one of them is not closed/destroyed during closing of the application. Probably this is the root cause. I will combine those two pools into one and hope the issue won't appear again.
Thanks,James

Date: Wed, 30 Mar 2016 21:17:23 -0700

**Comment by james-hu** [comment]
And, since it will take some time, I don't mind if the issue got closed for now. :-)

From: james.hu.ustc@hotmail.com
To: reply@reply.github.com; c3p0@noreply.github.com
Subject: RE: [swaldman/c3p0] CheckIdleResourcesTask.run() throws ClassNotFoundException when redeploying the application in Tomcat (#66)
Date: Thu, 31 Mar 2016 15:30:33 +1100




Hi Steve,
Thanks for the prompt reply. New finding to share: I created two pools and one of them is not closed/destroyed during closing of the application. Probably this is the root cause. I will combine those two pools into one and hope the issue won't appear again.
Thanks,James

Date: Wed, 30 Mar 2016 21:17:23 -0700
